client/components/webpack-build-monitor/index.jsx

```diff
@@ -0,0 +1,111 @@
+/* eslint-disable no-console */
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import { createStore } from 'redux';
+import classNames from 'classnames';
+import { find, includes, startsWith } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import Spinner from 'components/spinner';
+
+const CONNECTED = 'CONNECTED';
+const DISCONNECTED = 'DISCONNECTED';
+const NEEDS_RELOAD = 'NEEDS_RELOAD';
+const BUILDING_CSS = 'BUILDING_CSS';
+const BUILDING_JS = 'BUILDING_JS';
+const BUILDING_BOTH = 'BUILDING_BOTH';
+const ERROR = 'ERROR';
+const IDLE = 'IDLE';
+
+const doneBuilding = typeDone => state => {
+	if ( state === ERROR || state === NEEDS_RELOAD ) {
+		return state;
+	} else if ( state === BUILDING_BOTH ) {
+		return typeDone === BUILDING_JS ? BUILDING_CSS : BUILDING_JS;
+	}
+	return IDLE;
+};
+
+const buildJS = state => ( state === BUILDING_CSS ? BUILDING_BOTH : BUILDING_JS );
+const buildCSS = state => ( state === BUILDING_JS ? BUILDING_BOTH : BUILDING_CSS );
+
+const MESSAGE_STATUS_MAP = {
+	'[HMR] connected': () => CONNECTED,
+	'[WDS] Disconnected!': () => DISCONNECTED,
+	'[HMR] bundle rebuilding': state => buildJS( state ),
+	'Building CSS…': state => buildCSS( state ),
+	'CSS build failed': () => ERROR,
+	'[WDS] Nothing changed.': doneBuilding( BUILDING_JS ),
+	'Reloading CSS: ': doneBuilding( BUILDING_CSS ),
+	'[HMR] bundle rebuilt in': doneBuilding( BUILDING_JS ),
+	"[HMR] The following modules couldn't be hot updated": () => NEEDS_RELOAD,
+};
+
+const reducer = ( state = IDLE, { message } ) => {
+	const getNextState = find( MESSAGE_STATUS_MAP, ( v, messagePrefix ) =>
+		startsWith( message, messagePrefix ),
+	);
+
+	return getNextState ? getNextState( state ) : state;
+};
```

Comment:
This is probably good for now (and I'd love to start using it 😍 ). I was imagining something where you can still see if something is building or has errors, but rather than returning to `IDLE` when builds are not in progress, you return to `NEEDS_RELOAD`.

That might not be too relevant for JS, but I can imagine wanting to continue tweaking CSS and see build status even though you need to refresh for JS.